File: backends/java/zappy-application/src/test/java/com/zappymart/application/accounts/AccountUseCasesTest.java
package com.zappymart.application.accounts;

import com.zappymart.application.Visitor;
import com.zappymart.application.cart.AddToCart;
import com.zappymart.application.cart.CurrentCart;
import com.zappymart.application.fakes.TheStore;
import com.zappymart.domain.accounts.Session;
import com.zappymart.domain.catalogue.Product;
import com.zappymart.domain.shared.Result;
import com.zappymart.domain.shared.UserError;
import com.zappymart.domain.shared.UserErrorCode;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import static com.zappymart.application.fakes.SeedLikeData.JANE;
import static com.zappymart.application.fakes.SeedLikeData.SHIRT;
import static com.zappymart.application.fakes.SeedLikeData.SOLD_OUT_RING;
import static org.assertj.core.api.Assertions.assertThat;

class AccountUseCasesTest {

    private static final String THE_PASSWORD = "correct horse battery staple";

    private final TheStore store = new TheStore().holding(SHIRT, SOLD_OUT_RING).knowing(JANE);

    private final Visitor anonymous = new Visitor("cart-cookie", null, null, "curl", "http://localhost:5173");

    private CurrentCart currentCart;
    private SessionOpening sessionOpening;
    private AnonymousHandover anonymousHandover;
    private RegisterCustomer registerCustomer;
    private LogInCustomer logInCustomer;
    private RefreshSession refreshSession;
    private LogOut logOut;
    private RevokeSession revokeSession;
    private IdentifyVisitor identifyVisitor;
    private ViewWishlist viewWishlist;
    private AddToWishlist addToWishlist;
    private RemoveFromWishlist removeFromWishlist;

    @BeforeEach
    void wireTheUseCases() {
        currentCart = new CurrentCart(store.cartRepository, store.identifierGenerator, store.clock);
        sessionOpening = new SessionOpening(store.sessionStore, store.tokenIssuer, store.identifierGenerator,
                store.clock);
        anonymousHandover = new AnonymousHandover(currentCart, store.cartRepository, store.wishlistRepository,
                store.identifierGenerator, store.clock);
        registerCustomer = new RegisterCustomer(store.unitOfWork, store.customerRepository, store.passwordHasher,
                sessionOpening, anonymousHandover, store.rateLimiter, store.identifierGenerator, store.clock);
        logInCustomer = new LogInCustomer(store.unitOfWork, store.customerRepository, store.passwordHasher,
                sessionOpening, anonymousHandover, store.rateLimiter);
        refreshSession = new RefreshSession(store.unitOfWork, store.sessionStore, store.customerRepository,
                store.tokenIssuer, sessionOpening, store.clock);
        logOut = new LogOut(store.unitOfWork, store.sessionStore, store.tokenIssuer, store.clock);
        revokeSession = new RevokeSession(store.unitOfWork, store.sessionStore, store.clock);
        identifyVisitor = new IdentifyVisitor(store.tokenIssuer, store.sessionStore, store.identifierGenerator,
                store.clock);
        viewWishlist = new ViewWishlist(store.wishlistRepository, store.productRepository);
        addToWishlist = new AddToWishlist(store.unitOfWork, store.wishlistRepository, store.productRepository,
                viewWishlist);
        removeFromWishlist = new RemoveFromWishlist(store.unitOfWork, store.wishlistRepository, viewWishlist);
    }

    @Test
    void registersACustomerAndSignsThemInAtOnce() {
        Authentication authentication = registerCustomer
                .execute(anonymous, "New@Example.com", "New Customer", "a long enough password")
                .valueOrThrow();

        assertThat(authentication.customer().email().value()).isEqualTo("new@example.com");
        assertThat(authentication.accessToken().value()).startsWith("access:");
        assertThat(authentication.refreshToken()).isNotBlank();
        assertThat(store.sessions).containsKey(authentication.sessionId());
    }

    @Test
    void refusesARegistrationThatBreaksARule() {
        assertThat(reasonOf(registerCustomer.execute(anonymous, "nope", "Name", "a long enough password")))
                .isEqualTo(UserErrorCode.EMAIL_INVALID);
        assertThat(reasonOf(registerCustomer.execute(anonymous, "new@example.com", "Name", "short")))
                .isEqualTo(UserErrorCode.PASSWORD_TOO_SHORT);
        assertThat(reasonOf(registerCustomer.execute(anonymous, "jane@example.com", "Name", "a long password")))
                .isEqualTo(UserErrorCode.EMAIL_TAKEN);

        store.rateLimiterAllows = false;

        assertThat(reasonOf(registerCustomer.execute(anonymous, "other@example.com", "Name", "a long password")))
                .isEqualTo(UserErrorCode.RATE_LIMITED);
    }

    @Test
    void logsInAndAnswersOneReasonWhateverIsWrong() {
        assertThat(logInCustomer.execute(anonymous, "JANE@example.com", THE_PASSWORD, "Chrome")
                .valueOrThrow().customer().id()).isEqualTo("customer-01");
        assertThat(reasonOf(logInCustomer.execute(anonymous, "jane@example.com", "wrong password here", null)))
                .isEqualTo(UserErrorCode.CREDENTIALS_INVALID);
        assertThat(reasonOf(logInCustomer.execute(anonymous, "nobody@example.com", THE_PASSWORD, null)))
                .isEqualTo(UserErrorCode.CREDENTIALS_INVALID);
        assertThat(reasonOf(logInCustomer.execute(anonymous, "not an address", THE_PASSWORD, null)))
                .isEqualTo(UserErrorCode.CREDENTIALS_INVALID);
    }

    @Test
    void handsTheAnonymousCartAndWishlistToTheCustomerOnLogin() {
        new AddToCart(store.unitOfWork, currentCart, store.cartRepository, store.productRepository,
                store.identifierGenerator, store.clock).execute(anonymous, SHIRT.id(), 2);
        addToWishlist.execute(anonymous, SHIRT.id());

        Authentication authentication = logInCustomer
                .execute(anonymous, "jane@example.com", THE_PASSWORD, "Chrome").valueOrThrow();
        Visitor signedIn = new Visitor("cart-cookie", authentication.customer().id(),
                authentication.sessionId(), "curl", "http://localhost:5173");

        assertThat(currentCart.forVisitor(signedIn).lines()).hasSize(1);
        assertThat(viewWishlist.execute(signedIn)).extracting(Product::id).containsExactly(SHIRT.id());
    }

    @Test
    void rotatesTheRefreshTokenAndRevokesTheFamilyOnAReplay() {
        Authentication first = logInCustomer.execute(anonymous, "jane@example.com", THE_PASSWORD, "Chrome")
                .valueOrThrow();

        Authentication second = refreshSession.execute(first.refreshToken()).valueOrThrow();

        assertThat(second.refreshToken()).isNotEqualTo(first.refreshToken());
        assertThat(reasonOf(refreshSession.execute(first.refreshToken())))
                .isEqualTo(UserErrorCode.SESSION_INVALID);
        assertThat(reasonOf(refreshSession.execute(second.refreshToken())))
                .isEqualTo(UserErrorCode.SESSION_INVALID);
        assertThat(reasonOf(refreshSession.execute(null))).isEqualTo(UserErrorCode.SESSION_INVALID);
        assertThat(reasonOf(refreshSession.execute("a token nobody issued")))
                .isEqualTo(UserErrorCode.SESSION_INVALID);
    }

    @Test
    void aLoggedOutSessionIsRefusedAtOnceEvenWithAValidToken() {
        Authentication authentication = logInCustomer.execute(anonymous, "jane@example.com", THE_PASSWORD, "Chrome")
                .valueOrThrow();
        Visitor signedIn = identifyVisitor.execute(authentication.accessToken().value(), "cart-cookie", "curl",
                "http://localhost:5173");

        assertThat(signedIn.isSignedIn()).isTrue();

        logOut.execute(signedIn, authentication.refreshToken());

        assertThat(identifyVisitor.execute(authentication.accessToken().value(), "cart-cookie", "curl", null)
                .isSignedIn()).isFalse();
    }

    @Test
    void logsOutThroughTheRefreshCookieWhenNoAccessTokenCame() {
        Authentication authentication = logInCustomer.execute(anonymous, "jane@example.com", THE_PASSWORD, "Chrome")
                .valueOrThrow();

        assertThat(logOut.execute(anonymous, authentication.refreshToken())).isTrue();
        assertThat(logOut.execute(anonymous, null)).isTrue();
        assertThat(store.sessions.get(authentication.sessionId()).isOpenAt(TheStore.NOW)).isFalse();
    }

    @Test
    void revokesOneSessionOfTheSignedInCustomer() {
        Authentication first = logInCustomer.execute(anonymous, "jane@example.com", THE_PASSWORD, "Chrome")
                .valueOrThrow();
        Authentication second = logInCustomer.execute(anonymous, "jane@example.com", THE_PASSWORD, "Firefox")
                .valueOrThrow();
        Visitor signedIn = new Visitor("cart-cookie", "customer-01", second.sessionId(), "curl", null);

        List<Session> stillOpen = revokeSession.execute(signedIn, first.sessionId()).valueOrThrow();

        assertThat(stillOpen).extracting(Session::id).containsExactly(second.sessionId());
        assertThat(reasonOf(revokeSession.execute(signedIn, "session-that-is-not-there")))
                .isEqualTo(UserErrorCode.SESSION_NOT_FOUND);
        assertThat(reasonOf(revokeSession.execute(anonymous, first.sessionId())))
                .isEqualTo(UserErrorCode.NOT_AUTHENTICATED);
    }

    @Test
    void anAnonymousVisitorKeepsAWishlistAgainstTheCartCookie() {
        assertThat(addToWishlist.execute(anonymous, SOLD_OUT_RING.id()).valueOrThrow())
                .extracting(Product::id).containsExactly(SOLD_OUT_RING.id());
        assertThat(addToWishlist.execute(anonymous, SHIRT.id()).valueOrThrow())
                .extracting(Product::id).containsExactly(SHIRT.id(), SOLD_OUT_RING.id());
        assertThat(removeFromWishlist.execute(anonymous, SOLD_OUT_RING.id()).valueOrThrow())
                .extracting(Product::id).containsExactly(SHIRT.id());
        assertThat(reasonOf(addToWishlist.execute(anonymous, "product-99")))
                .isEqualTo(UserErrorCode.PRODUCT_NOT_FOUND);
    }

    @Test
    void findsTheSignedInCustomerAndNobodyElse() {
        FindSignedInCustomer findSignedInCustomer = new FindSignedInCustomer(store.customerRepository);
        Visitor signedIn = new Visitor("cart-cookie", "customer-01", "session-01", "curl", null);

        assertThat(findSignedInCustomer.execute(signedIn)).contains(JANE);
        assertThat(findSignedInCustomer.execute(anonymous)).isEmpty();
    }

    @Test
    void listsTheOpenSessionsOfACustomer() {
        logInCustomer.execute(anonymous, "jane@example.com", THE_PASSWORD, "Chrome");
        ListOpenSessions listOpenSessions = new ListOpenSessions(store.sessionStore, store.clock);

        assertThat(listOpenSessions.execute("customer-01")).hasSize(1);
    }

    private static <TValue> UserErrorCode reasonOf(Result<TValue> result) {
        return result.errors().stream().map(UserError::code).findFirst().orElseThrow();
    }
}

File: backends/kotlin/zappy-application/src/test/kotlin/nl/zappymart/application/accounts/AccountUseCasesTest.kt
package nl.zappymart.application.accounts

import java.time.Instant
import nl.zappymart.application.AllowingRateLimiter
import nl.zappymart.application.CountingIdentifierFactory
import nl.zappymart.application.DirectUnitOfWork
import nl.zappymart.application.FixedClock
import nl.zappymart.application.InMemoryCarts
import nl.zappymart.application.InMemoryCustomers
import nl.zappymart.application.InMemoryProducts
import nl.zappymart.application.InMemoryPromotions
import nl.zappymart.application.InMemorySessions
import nl.zappymart.application.InMemoryWishlists
import nl.zappymart.application.PredictableAccessTokenIssuer
import nl.zappymart.application.PredictableRefreshTokenIssuer
import nl.zappymart.application.ReversingPasswordHasher
import nl.zappymart.application.Visitor
import nl.zappymart.application.cart.AddToCart
import nl.zappymart.application.cart.CartPromotion
import nl.zappymart.application.cart.VisitorCart
import nl.zappymart.domain.accounts.Customer
import nl.zappymart.domain.accounts.PasswordHash
import nl.zappymart.domain.builders.aProduct
import nl.zappymart.domain.shared.EmailAddress
import nl.zappymart.domain.shared.Result
import nl.zappymart.domain.shared.UserErrorCode
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class AccountUseCasesTest {

    private val jacket = aProduct().withId("product-03").named("Cotton jacket").costing(5599).withStock(8).build()

    private val clock = FixedClock()

    private val passwords = ReversingPasswordHasher()

    private val jane = Customer(
        "customer-01",
        EmailAddress.ofStored("jane@example.com"),
        "Jane Doe",
        passwords.hash("correct horse battery staple"),
        Instant.parse("2026-01-15T09:00:00Z"),
    )

    private val customers = InMemoryCustomers(listOf(jane))

    private val sessions = InMemorySessions()

    private val carts = InMemoryCarts()

    private val wishlists = InMemoryWishlists()

    private val products = InMemoryProducts(listOf(jacket))

    private val identifiers = CountingIdentifierFactory()

    private val unitOfWork = DirectUnitOfWork()

    private val accessTokens = PredictableAccessTokenIssuer()

    private val refreshTokens = PredictableRefreshTokenIssuer()

    private val rateLimiter = AllowingRateLimiter()

    private val visitorCart = VisitorCart(carts, clock, identifiers)

    private val cartPromotion = CartPromotion(InMemoryPromotions(), clock)

    private val addToCart = AddToCart(visitorCart, cartPromotion, carts, products, identifiers, clock, unitOfWork)

    private val wishlistOwner = WishlistOwner(wishlists, visitorCart, carts)

    private val viewWishlist = ViewWishlist(wishlists, products, wishlistOwner)

    private val addToWishlist = AddToWishlist(wishlists, products, viewWishlist, wishlistOwner, unitOfWork)

    private val removeFromWishlist = RemoveFromWishlist(wishlists, viewWishlist, wishlistOwner, unitOfWork)

    private val signIn = SignIn(sessions, accessTokens, refreshTokens, identifiers, clock)

    private val logInCustomer =
        LogInCustomer(customers, passwords, signIn, visitorCart, wishlistOwner, rateLimiter, unitOfWork)

    private val registerCustomer = RegisterCustomer(
        customers,
        passwords,
        identifiers,
        clock,
        signIn,
        visitorCart,
        wishlistOwner,
        rateLimiter,
        unitOfWork,
    )

    private val refreshSession =
        RefreshSession(sessions, customers, accessTokens, refreshTokens, clock, unitOfWork)

    private val logOut = LogOut(sessions, refreshTokens, clock, unitOfWork)

    private val revokeSession = RevokeSession(sessions, clock, unitOfWork)

    private val identifyVisitor = IdentifyVisitor(accessTokens, sessions, clock)

    @Test
    fun `a login with the seed password opens a session`() {
        val signedIn = logInCustomer.execute(
            Visitor.ANONYMOUS,
            LoginRequest("JANE@Example.com", "correct horse battery staple", "Chrome on Windows"),
        ) as Result.Success
        assertThat(signedIn.value.customer.id).isEqualTo("customer-01")
        assertThat(sessions.findById(signedIn.value.sessionId)?.device).isEqualTo("Chrome on Windows")
    }

    @Test
    fun `a wrong password answers one code and never says whether the address exists`() {
        val wrongPassword = logInCustomer.execute(
            Visitor.ANONYMOUS,
            LoginRequest("jane@example.com", "not the right password", "Chrome"),
        ) as Result.Refused
        val unknownAddress = logInCustomer.execute(
            Visitor.ANONYMOUS,
            LoginRequest("nobody@example.com", "not the right password", "Chrome"),
        ) as Result.Refused
        assertThat(wrongPassword.errors).isEqualTo(unknownAddress.errors)
        assertThat(wrongPassword.errors.map { error -> error.code })
            .containsExactly(UserErrorCode.CREDENTIALS_INVALID)
    }

    @Test
    fun `too many attempts are rate limited`() {
        rateLimiter.refusesEverything()
        val refused = logInCustomer.execute(
            Visitor.ANONYMOUS,
            LoginRequest("jane@example.com", "correct horse battery staple", "Chrome"),
        ) as Result.Refused
        assertThat(refused.errors.map { error -> error.code }).containsExactly(UserErrorCode.RATE_LIMITED)
    }

    @Test
    fun `registering with a taken address is refused`() {
        val refused = registerCustomer.execute(
            Visitor.ANONYMOUS,
            RegistrationRequest("jane@example.com", "Jane", "a long enough password", "Chrome"),
        ) as Result.Refused
        assertThat(refused.errors.map { error -> error.code }).containsExactly(UserErrorCode.EMAIL_TAKEN)
    }

    @Test
    fun `registering signs the new customer in and takes the cart along`() {
        val started = addToCart.execute(Visitor.ANONYMOUS, "product-03", 1)
        val visitor = Visitor(null, null, started.cart.id)
        val registered = registerCustomer.execute(
            visitor,
            RegistrationRequest("  NEW@Example.com ", "New Customer", "a long enough password", "Chrome"),
        ) as Result.Success
        assertThat(registered.value.customer.email.value).isEqualTo("new@example.com")
        assertThat(carts.findByCustomerId(registered.value.customer.id)?.lines).hasSize(1)
    }

    @Test
    fun `a refresh token is used once and a replay revokes the session`() {
        val signedIn = logInCustomer.execute(
            Visitor.ANONYMOUS,
            LoginRequest("jane@example.com", "correct horse battery staple", "Chrome"),
        ) as Result.Success
        val firstToken = signedIn.value.refreshToken

        val refreshed = refreshSession.execute(firstToken) as Result.Success
        assertThat(refreshed.value.refreshToken).isNotEqualTo(firstToken)

        val replayed = refreshSession.execute(firstToken) as Result.Refused
        assertThat(replayed.errors.map { error -> error.code }).containsExactly(UserErrorCode.SESSION_INVALID)

        val afterTheReplay = refreshSession.execute(refreshed.value.refreshToken) as Result.Refused
        assertThat(afterTheReplay.errors.map { error -> error.code }).containsExactly(UserErrorCode.SESSION_INVALID)
    }

    @Test
    fun `a refresh without a token is refused`() {
        val refused = refreshSession.execute(null) as Result.Refused
        assertThat(refused.errors.map { error -> error.code }).containsExactly(UserErrorCode.SESSION_INVALID)
    }

    @Test
    fun `a logged out session is refused on the next bearer request`() {
        val signedIn = logInCustomer.execute(
            Visitor.ANONYMOUS,
            LoginRequest("jane@example.com", "correct horse battery staple", "Chrome"),
        ) as Result.Success
        val bearer = signedIn.value.accessToken.value
        assertThat(identifyVisitor.execute(bearer, null).customerId).isEqualTo("customer-01")

        logOut.execute(Visitor("customer-01", signedIn.value.sessionId, null), signedIn.value.refreshToken)
        assertThat(identifyVisitor.execute(bearer, null).customerId).isNull()
    }

    @Test
    fun `logging out twice answers the same`() {
        val signedIn = logInCustomer.execute(
            Visitor.ANONYMOUS,
            LoginRequest("jane@example.com", "correct horse battery staple", "Chrome"),
        ) as Result.Success
        val visitor = Visitor("customer-01", signedIn.value.sessionId, null)
        assertThat(logOut.execute(visitor, signedIn.value.refreshToken)).isTrue()
        assertThat(logOut.execute(visitor, signedIn.value.refreshToken)).isTrue()
    }

    @Test
    fun `revoking a session that belongs to somebody else is refused`() {
        val signedIn = logInCustomer.execute(
            Visitor.ANONYMOUS,
            LoginRequest("jane@example.com", "correct horse battery staple", "Chrome"),
        ) as Result.Success
        val refused = revokeSession.execute(
            Visitor("customer-01", signedIn.value.sessionId, null),
            "a-session-of-somebody-else",
        ) as Result.Refused
        assertThat(refused.errors.map { error -> error.code }).containsExactly(UserErrorCode.SESSION_NOT_FOUND)
    }

    @Test
    fun `an anonymous wishlist lives against the cart cookie and merges on login`() {
        val saved = addToWishlist.execute(Visitor.ANONYMOUS, "product-03")
        assertThat(saved.errors).isEmpty()
        assertThat(saved.products.map { product -> product.id }).containsExactly("product-03")
        val anonymousCartId = requireNotNull(saved.anonymousCartId)

        val visitor = Visitor(null, null, anonymousCartId)
        assertThat(viewWishlist.execute(visitor).map { product -> product.id }).containsExactly("product-03")

        logInCustomer.execute(visitor, LoginRequest("jane@example.com", "correct horse battery staple", "Chrome"))
        val asCustomer = Visitor("customer-01", "session-01", anonymousCartId)
        assertThat(viewWishlist.execute(asCustomer).map { product -> product.id }).containsExactly("product-03")
    }

    @Test
    fun `saving an unknown product to a wishlist is refused`() {
        val refused = addToWishlist.execute(Visitor.ANONYMOUS, "product-99")
        assertThat(refused.errors.map { error -> error.code }).containsExactly(UserErrorCode.PRODUCT_NOT_FOUND)
    }

    @Test
    fun `removing a product that is not on the wishlist is not an error`() {
        val change = removeFromWishlist.execute(Visitor.ANONYMOUS, "product-03")
        assertThat(change.errors).isEmpty()
        assertThat(change.products).isEmpty()
    }

    @Test
    fun `an unknown bearer token leaves the visitor anonymous`() {
        assertThat(identifyVisitor.execute("not-a-token", "cart-01"))
            .isEqualTo(Visitor(null, null, "cart-01"))
        assertThat(FindSignedInCustomer(customers).execute(Visitor.ANONYMOUS)).isNull()
        assertThat(ListSessions(sessions, clock).execute("customer-01")).isEmpty()
    }
}
